app/routes/task.py

```diff
@@ -0,0 +1,143 @@
+
+#constants: (refactor to make this from the database later.)
+COL_NAMES = ["title", "description", "is_complete"] #later, add completed ad
+COL_DEFAULTS = [None, "", False]
+COL_NAME_DEFAULT_DICT = dict(zip(COL_NAMES, COL_DEFAULTS))
+
+@task_bp.route("", methods = ["GET"])
+def get_all_tasks():
+    #refactor to helper function
+    sort_order = request.args.get("sort")
+    if sort_order == "desc":
+        tasks = Task.query.order_by(Task.title.desc()).all() #Task.title is just a string??
+    elif sort_order == "asc":
+        tasks = Task.query.order_by(Task.title.asc()).all() #look up doc for asc.
+    else:
+        tasks = Task.query.all()
+    #refactor to helper function end here
+    response = []
+    for task in tasks:
+        task_dict = task.make_dict()
+        response.append(task_dict)
+    return jsonify(response), 200
+
+
+@task_bp.route("/<task_id>", methods = ["GET"])
+def get_one_task(task_id):
+    task = validate_task(task_id)
+    task_dict = task.make_dict()
+    return make_response({"task": task_dict}, 200)
+    
+@task_bp.route("", methods = ["POST"])
+def post_new_task():
+    request_body = request.get_json()
+    #dict_of_field_vals = fill_empties_with_defaults(request_body)
+    if "title" not in request_body or "description" not in request_body:
+        response_str = "Invalid data"
+        abort(make_response({"details":response_str}, 400))
+    if "completed_at" not in request_body:
+        request_body["is_complete"] = False
+
+    new_task = Task.from_dict(request_body)
+    db.session.add(new_task)
+    db.session.commit()
+    task_dict = new_task.make_dict()
+    response = {"task": task_dict}
+    return make_response(response, 201)
+    
+@task_bp.route("/<task_id>", methods = ["PUT", "PATCH"])
+def update_task(task_id):
+    task = validate_task(task_id)
+    request_body = request.get_json()
+    task = update_given_values(task, request_body)
+    db.session.commit()
+    response = {"task": task.make_dict()}  #refactor this line and line 37 above to helper function? or method on class?
+    return make_response(response, 200)
+
+#can the following be combined with the route above?
+@task_bp.route("/<task_id>/<complete_tag>", methods = ["PUT", "PATCH"])
+def update_task_completion(task_id, complete_tag):
+    task = validate_task(task_id)
+    if complete_tag == "mark_complete":
+        task.completed_at = date.today().strftime("%B %d, %Y")
+        task.is_complete = True  #not sure if this line is redundant
+
+        #post a message to slack to say the task is complete.
+        #----------Make the following into a helper function later-----#
+        path = "https://slack.com/api/chat.postMessage"
+        query_params = {
+            "channel" : "task-notifications",
+            "text": "Someone just completed the task " + task.title
+        }
+        headers = {
+            "Authorization" : "Bearer " + os.environ.get("SLACK_API_KEY")
```

Comment:
The reason Learn is failing is that the Environment variable used by Learn is not named SLACK_API_KEY and so `os.environ.get("SLACK_API_KEY")` returns `None`.